### ISSUE TITLE
Memory Allocation and Performance Considerations

### ISSUE THREAD

**Issue by laino** [pixijs/pixi.js#6332]
**tl;dr:** Get 600% faster code with this one simple trick!

So I was looking into memory allocation patterns of my application and looking for stuff that might cause GC pauses, which lead me to wonder about a pattern that is quite common in this library.

Right now most internal state is kept in arrays that are routinely cleared
with ``array.length = 0``, their old elements being pushed into some kind of pool,
then later retrieved from there and added to the array again.

The idea here is to reduce memory allocation/GC pauses in an attempt to improve performance. The problem with that approach is that it misses the real elephant in the room, which is that doing

```js
arr.length = 0; // memory freed!
arr.push(X); // memory allocation!
```

is not only slow, it also causes memory to be freed and allocated.
A class where this happens a lot is ``GraphicsGeometry``, which is the culprit that actually made me investigate this in the first place.

How bad is it? Let's run some benchmarks.

Consider this code that roughly mimics the way ``GraphicsGeometry`` works:

```js
const {performance} = require('perf_hooks');

const DATA_STORE_POOL = [];

class DataStore {
    constructor() {
        this.x = 0;
        this.y = 0;
    }
}

function RENDER_DATA(arr) {
    // Fake render. In our case 'rendering' is summing the array.
    let sum = 0;

    for (let i = 0; i < arr.length; i++) {
        sum += arr[i].x + arr[i].y;
    }

    return sum;
}

class AllocationPatternCurrent {
    constructor() {
        this.internalArray = [];
    }

    addData(x, y) {
        const store = DATA_STORE_POOL.pop() || new DataStore();

        store.x = x;
        store.y = y;

        // This will cause an allocation
        this.internalArray.push(store);
    }

    clear() {
        const internalArray = this.internalArray;

        for (let i = 0; i < internalArray.length; i++) {
            // This will also cause an allocation
            DATA_STORE_POOL.push(internalArray[i]);
        }

        internalArray.length = 0;
    }

    render() {
        return RENDER_DATA(this.internalArray);
    }

    destroy() {
        this.internalArray.length = 0;
        this.dataCount = 0;
    }
}

function testPattern(Pattern) {
    const store = new Pattern();

    let renderSum = 0;

    for (let i = 0; i < 1000000; i++) {
        store.clear();

        // Some shape with 4 points
        store.addData(10 + (i % 10), 5);
        store.addData(10, 10 + (i % 100));
        store.addData(5 - (i % 1000), 10);
        store.addData(5, 5 - (i % 10000));

        // Ocassionally add another point
        if (i % 10 === 0) {
            store.addData(1, 0 - (i % 3));
        }

        renderSum += store.render();
    }

    return renderSum;
}

for (let i = 0; i < 10; i++) {
    const start = performance.now();
    const result = testPattern(AllocationPatternCurrent);
    const end = performance.now();

    console.log("Took:", end - start, "Result:", result);
}
```

Let's run it:

```
Took: 161.43071299791336 Result: -5384999999
Took: 155.63587003946304 Result: -5384999999
Took: 160.9160519838333 Result: -5384999999
Took: 198.32747304439545 Result: -5384999999
Took: 184.14101201295853 Result: -5384999999
Took: 206.88946896791458 Result: -5384999999
Took: 219.52886599302292 Result: -5384999999
Took: 195.83152198791504 Result: -5384999999
Took: 183.58255696296692 Result: -5384999999
Took: 183.1462979912758 Result: -5384999999
```

Memory use was constant all the while, as expected, since we're using an object pool. Right?

But can we do better? Most of the time after clearing we add the same amount of data again before rendering, since we only changed the shape of our imaginary polygon. So what if we only changed the length of our array lazily and used an internal counter to keep track of the real array length?

Consider this code:

```js
const {performance} = require('perf_hooks');

class DataStore {
    constructor() {
        this.x = 0;
        this.y = 0;
    }
}

function RENDER_DATA(arr) {
    // Fake render. In our case 'rendering' is summing the array.
    let sum = 0;

    for (let i = 0; i < arr.length; i++) {
        sum += arr[i].x + arr[i].y;
    }

    return sum;
}

class AllocationPatternProposed {
    constructor() {
        this.dataCount = 0;
        this.internalArray = [];
    }

    addData(x, y) {
        const internalArray = this.internalArray;

        let store;

        if (internalArray.length > this.dataCount) {
            store = this.internalArray[this.dataCount];
        } else {
            store = new DataStore();
            this.internalArray.push(store);
        }

        store.x = x;
        store.y = y;

        this.dataCount++;
    }

    clear() {
        this.dataCount = 0;
    }

    render() {
        // Finalize internal data
        const internalArray = this.internalArray;
        const dataCount = this.dataCount;

        if (internalArray.length !== dataCount) {
            internalArray.length = dataCount;
        }

        return RENDER_DATA(internalArray);
    }

    destroy() {
        this.internalArray.length = 0;
    }
}

function testPattern(Pattern) {
    const store = new Pattern();

    let renderSum = 0;

    for (let i = 0; i < 1000000; i++) {
        store.clear();

        // Some shape with 4 points
        store.addData(10 + (i % 10), 5);
        store.addData(10, 10 + (i % 100));
        store.addData(5 - (i % 1000), 10);
        store.addData(5, 5 - (i % 10000));

        // Ocassionally add another point
        if (i % 10 === 0) {
            store.addData(1, 0 - (i % 3));
        }

        renderSum += store.render();
    }

    return renderSum;
}

for (let i = 0; i < 10; i++) {
    const start = performance.now();
    const result = testPattern(AllocationPatternProposed);
    const end = performance.now();

    console.log("Took:", end - start, "Result:", result);
}
```

Here we have an internal counter called ``dataCount`` that keeps track of how many elements are in our array, and we only ``.push()`` if there is not enough space left in our current array. That means if
just add a shape with the same number of points after a ``.clear()``, we are actually completely allocation free.

You might also notice that I got rid of the data pool, though both patterns could be trivially combined (but let's face it: the JavaScript engine is probably better at pooling objects with just 2 number properties).

Let's run this thing:
```
Took: 51.44574695825577 Result: -5384999999
Took: 38.26657301187515 Result: -5384999999
Took: 41.524843990802765 Result: -5384999999
Took: 38.86036801338196 Result: -5384999999
Took: 31.547271013259888 Result: -5384999999
Took: 31.820671021938324 Result: -5384999999
Took: 31.354419946670532 Result: -5384999999
Took: 31.34599995613098 Result: -5384999999
Took: 31.31634098291397 Result: -5384999999
Took: 31.361370027065277 Result: -5384999999
```

This was a whopping **600%** faster and also runs in constant memory.

Also note how the measured time becomes almost perfectly constant once the engine had a chance to optimize it, unlike the other example's runtime which is... interesting. If you increase the loop iterations both should reach a more-or-less constant runtime though.

Both tests were done on node v13.6.0.

You can run these tests in your browser by omitting the ``require`` line and copy & pasting them into the browser console. On my Firefox the difference is 1100ms vs. 380ms.

**Edit:** I should add that with *allocation* I mean allocated from the heap and with *constant memory* I am speaking of the heap size being constant.

**Comment by ivanpopelyshev** [comment]
```js
const DATA_STORE_POOL = [];

class DataStore {
    constructor() {
        this.x = 0;
        this.y = 0;
    }
}

function RENDER_DATA(arr, cnt) {
    // Fake render. In our case 'rendering' is summing the array.
    let sum = 0;

    for (let i = 0; i < cnt; i++) {
        sum += arr[i].x + arr[i].y;
    }

    return sum;
}

class AllocationPatternCurrent {
    constructor() {
        this.internalArray = [];
		this.cnt = 0;
    }

    addData(x, y) {
        const store = DATA_STORE_POOL.pop() || new DataStore();

        store.x = x;
        store.y = y;

        // This will cause an allocation
        this.internalArray[this.cnt++] = store;
    }

    clear() {
        const internalArray = this.internalArray;

        for (let i = 0; i < this.cnt; i++) {
            // This will also cause an allocation
            DATA_STORE_POOL.push(internalArray[i]);
			internalArray[i] = null;
        }

        this.cnt = 0;
    }

    render() {
        return RENDER_DATA(this.internalArray, this.cnt);
    }

    destroy() {
        this.internalArray.length = 0;
        this.dataCount = 0;
    }
}
```

That's 40-80ms.

If we save internalArray lenght but just empty it - we do it in BatchRenderer.

We might add it in `GraphicsGeometry` batches.
However we still need that pool, because it exists for all `GraphicsGeometry` instances.

**Comment by ivanpopelyshev** [comment]
Also you are measuring cost PER REFILL , on 100k refills. Refill in graphics has many array allocations that certainly more costly than this thing.

In general I want refill Graphics to be faster, however fixing that particular thing wont affect it.

**Comment by laino** [comment]
In my case the allocations via ``.push`` in ``addUvs``, ``addTextureIds``, and ``updateBatches`` are the single worst memory bottlenecks per frame.

The ``GraphicsGeometry`` is a lot to take in though, so I started with this experiment before moving on  to take on the actual thing in a PR.

The reason I didn't go for what you proposed is that I wasn't sure what places would need to be changed if ``array.length !== dataCount`` and because I feel we should return some memory if the graphics object becomes significantly smaller again. The latter might be pointless though and the other one is just a lack of familiarity.

**Comment by ivanpopelyshev** [comment]
Oh those arrays eat the most memory, together with `points`, `shape.points`. Also they are duplicated in Buffers (maybe) because of typed array copies.

In case of less data, short them down only if its needed. That will help in case someone refills them often.

The real cases when its important are SVG and SWF formats. Morphing (animating) shapes is even bigger problem.

I honestly don't know whether its good to fix them part-by-part or to rewrite it all once more.

**Comment by laino** [comment]
> If we save internalArray lenght but just empty it - we do it in BatchRenderer.
> 
> We might add it in `GraphicsGeometry` batches.
> However we still need that pool, because it exists for all `GraphicsGeometry` instances.

I'm not sure what you mean by this.

Anyways. If we really want to use a pool, here's an idea to make that fast anyways.
First, here's the runtime of your example on my machine:

```
Took: 66.64429903030396 Result: -5384999999
Took: 50.39274698495865 Result: -5384999999
Took: 74.67829102277756 Result: -5384999999
Took: 71.28717398643494 Result: -5384999999
Took: 71.77907103300095 Result: -5384999999
Took: 71.45296800136566 Result: -5384999999
Took: 71.25986099243164 Result: -5384999999
Took: 71.02417999505997 Result: -5384999999
Took: 71.2890630364418 Result: -5384999999
Took: 71.13881701231003 Result: -5384999999
```

Now consider this slightly modified version:

```js
const {performance} = require('perf_hooks');

class GenericPool {
    constructor() {
        this.data = [];
        this.count = 0;
    }

    pop() {
        if (this.count) {
            this.count--;
            return this.data[this.count];
        }

        return null;
    }

    push(elem) {
        this.data[this.count++] = elem;
    }
}

const DATA_STORE_POOL = new GenericPool();

class DataStore {
    constructor() {
        this.x = 0;
        this.y = 0;
    }
}

function RENDER_DATA(arr) {
    // Fake render. In our case 'rendering' is summing the array.
    let sum = 0;

    for (let i = 0; i < arr.length; i++) {
        sum += arr[i].x + arr[i].y;
    }

    return sum;
}

class AllocationPatternProposed2 {
    constructor() {
        this.internalArray = [];
        this.cnt = 0;
    }

    addData(x, y) {
        const {internalArray, cnt} = this;

        let store;

        if (cnt < internalArray.length) {
            store = internalArray[cnt];
        } else {
            // This will cause an allocation
            internalArray.push(store = DATA_STORE_POOL.pop() || new DataStore());
        }

        store.x = x;
        store.y = y;

        this.cnt++;
    }

    clear() {
        this.cnt = 0;
    }

    render() {
        const {cnt, internalArray} = this;
        const len = internalArray.length;

        if (len !== cnt) {
            for (let i = cnt; i < len; i++) {
                DATA_STORE_POOL.push(internalArray[i]);
            }

            internalArray.length = cnt;
        }

        return RENDER_DATA(internalArray);
    }

    destroy() {
        this.internalArray.length = 0;
        this.dataCount = 0;
    }
}

function testPattern(Pattern) {
    const store = new Pattern();

    let renderSum = 0;

    for (let i = 0; i < 1000000; i++) {
        store.clear();

        // Some shape with 4 points
        store.addData(10 + (i % 10), 5);
        store.addData(10, 10 + (i % 100));
        store.addData(5 - (i % 1000), 10);
        store.addData(5, 5 - (i % 10000));

        // Ocassionally add another point
        if (i % 10 === 0) {
            store.addData(1, 0 - (i % 3));
        }

        renderSum += store.render();
    }

    return renderSum;
}

for (let i = 0; i < 10; i++) {
    const start = performance.now();
    const result = testPattern(AllocationPatternProposed2);
    const end = performance.now();

    console.log("Took:", end - start, "Result:", result);
}
```

Here we put unused objects into the pool on ``render``.

Running it:

```
Took: 51.48372399806976 Result: -5384999999
Took: 34.433080971241 Result: -5384999999
Took: 39.81157797574997 Result: -5384999999
Took: 32.416814029216766 Result: -5384999999
Took: 32.73268097639084 Result: -5384999999
Took: 32.54303401708603 Result: -5384999999
Took: 32.417153000831604 Result: -5384999999
Took: 32.588356018066406 Result: -5384999999
Took: 32.765583992004395 Result: -5384999999
Took: 32.55689096450806 Result: -5384999999
```

**Comment by ivanpopelyshev** [comment]
Good one
